PEP 440 Direct Reference support

FlorentJeannot opened the pull request:
This PR adds [PEP 440 Direct References](https://www.python.org/dev/peps/pep-0440/#direct-references) support to fix https://github.com/jazzband/pip-tools/issues/1391. 

`pip` can support direct references, so the egg can be omitted inside the URL (see [the code from pip](https://github.com/pypa/pip/blob/7ec0fa5142466f8402bb866a1dcfb8ca6ffdc66b/src/pip/_vendor/packaging/requirements.py#L149)).

In my projects which have git repositories specified as dependencies, without a direct reference or an egg, pip was not able to determine the package name, which was causing an issue during `pip-sync` (inside the merge, comparing two identical requirements from two different files).

Also, inside `setup.py`, `install_requires` must be written with a direct reference (`pip` fails to install the dependencies if it is not using a direct reference). `pip-tools` was ignoring the direct reference, so it was not generating the `requirements.txt` correctly. See my example:

```
# setup.py
install_requires=["package @ git+ssh://url@version"]

# requirements.txt (from pip-tools)
git+ssh://url@version
```

**Changelog-friendly one-liner**: PEP 440 Direct References support 

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

FlorentJeannot commented:
FYI I installed `pip-tools` in my project using my fork and I generated and compared the `requirements.txt` and `dev-requirements.txt` files with the new ones:

- there was no regression
- the direct references were generated when it was only needed
- one dependency of a dependency generated a URL with an egg, without direct reference (which is fine: it keeps the original behavior of the tool)

And `pip-sync` worked (it was not without this fix; see https://github.com/jazzband/pip-tools/issues/1391)

FlorentJeannot commented:
To help you understand this fix/enhancement, from the master branch, try the following tests:

### Test 1

In `pip-tools/tests/test_data/packages/fake_with_deps/setup.py`, add in `install_requires`:
```python
"requests @ git+git://github.com/psf/requests@v2.25.1"
```

In `test_realistic_complex_sub_dependencies`, add the following at the end:
```python
print(out.stderr)
assert "requests @ git+git://github.com/psf/requests@v2.25.1" in out.stderr
```

This test will fail because it is missing the direct reference, inside the output there is:
```
git+git://github.com/psf/requests@v2.25.1
```

Also, note that there is no `egg` specified.

### Test 2

In `pip-tools/tests/test_resolver.py`, add the following:

```python
def test_combine_install_requirements_direct_ref(repository, from_line):
    first_dep = from_line("git+ssh://****@bitbucket.org/private-bucket/enums.git@4.0.0", comes_from="-r requirements.in")
    second_dep = from_line("git+ssh://****@bitbucket.org/private-bucket/enums.git@4.0.0", comes_from="-r dev-requirements.in")

    combined = combine_install_requirements(repository, [first_dep, second_dep])
    assert combined.comes_from == first_dep.comes_from  # shortest string
    assert set(combined._source_ireqs) == {first_dep, second_dep}
    assert str(combined.req.specifier) == ""
```

This is failing because both names are missing a specifier. You can make the test pass by specifying a direct reference or an egg:

- `enums @ git+ssh://****@bitbucket.org/private-bucket/enums.git@4.0.0`
- `git+ssh://****@bitbucket.org/private-bucket/enums.git@4.0.0#egg=enums`

For example this works:

```python
def test_combine_install_requirements_direct_ref(repository, from_line):
    first_dep = from_line("enums @ git+ssh://****@bitbucket.org/private-bucket/enums.git@4.0.0", comes_from="-r requirements.in")
    second_dep = from_line("git+ssh://****@bitbucket.org/private-bucket/enums.git@4.0.0#egg=enums", comes_from="-r dev-requirements.in")

    combined = combine_install_requirements(repository, [first_dep, second_dep])
    assert combined.comes_from == first_dep.comes_from  # shortest string
    assert set(combined._source_ireqs) == {first_dep, second_dep}
    assert str(combined.req.specifier) == ""
```

### Conclusion

- Test 1 does not generate the requirement with the direct reference, even though it was specified in `install_requires `
- Test 2 fails if a specifier (a direct reference or an egg) is missing

Because `pip-tools` does not handle direct references, when:

- `requirements.in` contains a direct reference
- `dev-requirements.in` contains the same direct reference

Then the resolver fails when entering `combine_install_requirements`.

I hope this helps to make it clearer and I am looking forward to reading your feedback!

FlorentJeannot commented on tests/test_data/packages/fake_with_deps/setup.py:
```diff
@@ -18,5 +18,6 @@
         "SQLAlchemy!=0.9.5,<2.0.0,>=0.7.8,>=1.0.0",
         "python-memcached>=1.57,<2.0",
         "xmltodict<=0.11,>=0.4.6",
+        "requests @ git+git://github.com/psf/requests@v2.25.1",
```
This new dependency does not bring anything more by itself, so I could either:

1. remove it
2. assert inside `test_realistic_complex_sub_dependencies` that `out.stderr` contains a direct reference

AndydeCleyre commented:
Hey, you reached out last week over at #1329, and today I was able to start looking at this. It's *possible* that, aside from changes to tests, all that's needed on top of my WIP PR is:

```diff
diff --git a/piptools/utils.py b/piptools/utils.py
index 4f15729..2f81af1 100644
--- a/piptools/utils.py
+++ b/piptools/utils.py
@@ -1,6 +1,7 @@
 import collections
 import itertools
 import os
+import re
 import shlex
 from contextlib import contextmanager, suppress
 from typing import (
@@ -137,6 +138,8 @@ def format_requirement(
                         ireq.local_file_path, from_dir
                     ).replace(os.path.sep, "/")
             ireq_url = f"{path_url}{fragment_string(ireq)}"
+        if not re.match(r"^#(|.*&)egg=.*", fragment_string(ireq)) and ireq.name:
+            ireq_url = f"{ireq.name} @ {ireq_url}"
         line = f"-e {ireq_url}" if ireq.editable else ireq_url
 
     if marker:
```
I'll leave this diff right here, and if #1329 is ever done this will help me merge.

FlorentJeannot commented:
@AndydeCleyre Thanks for your suggestion. Yeah it looks like that would work, the tests I've added will help you find out. :)

luzfcb commented:
@FlorentJeannot I guess that's necessary to include these two test cases to ensure this continues to work in the future. Anyway looks good to me.

The output should be the same as the input:

```
django-debug-toolbar@https://github.com/jazzband/django-debug-toolbar/archive/6ec6bfe0ba5912d7fd0a2b98176a215847157696.zip

django-taggit@https://github.com/jazzband/django-taggit/archive/74460fdc4a8b8bf2d36fb8a59691403cde7b4528.zip#sha1=594b7dd32bec37d8bf70a6ffa8866d30e93f3c42

```

FlorentJeannot commented:
Hey @luzfcb, this PR would definitely fix your issue! :)

Concerning your suggestion, I think the existing tests are enough. Could you explain your point of view?

@atugushev Do you know when this PR will be reviewed?

Thanks!

atugushev commented on piptools/utils.py:
```diff
@@ -115,7 +115,16 @@ def format_requirement(
-        line = ireq.link.url
+        if not ireq.name:
+            line = ireq.link.url
+        elif ireq.link.url.startswith("file:./"):
```
How `file:.` would work?

FlorentJeannot replied:
This is to preserve an existing "hack" that allows us to specify a relative path, as mentioned [here](https://github.com/jazzband/pip-tools/issues/204#issuecomment-550051424).
Although, I can see this is not good enough: if the relative path does not start with `"./"`, such as `file:src/foo#egg=foo` then this does not work.
I'm going to update this.

FlorentJeannot replied:
The logic has been updated, so I am resolving this conversation if you don't mind.

atugushev commented on piptools/utils.py:
```diff
@@ -115,7 +115,18 @@ def format_requirement(
+        # if the requirement has no name then it's just a URL
+        if not ireq.name:
+            line = ireq.link.url
+        # if it starts with "file:", PEP508 does not support direct reference
+        elif ireq.link.url.startswith("file:"):
```
> PEP508 does not support direct reference

I couldn't find any information about that in the PEP. I've checked the grammar and couldn't found any restrictions on `name @ file:...` too.

I've tried locally absolute paths and it works:

```console
$ pip install 'pip-tools @ file:///Users/albert/Projects/pip-tools'
...
Installing collected packages: pip-tools
  Attempting uninstall: pip-tools
    Found existing installation: pip-tools 6.1.1.dev72+g39856be
    Uninstalling pip-tools-6.1.1.dev72+g39856be:
      Successfully uninstalled pip-tools-6.1.1.dev72+g39856be
Successfully installed pip-tools-6.1.1.dev72+g39856be
```

Local path doesn't work though:

```
$ pip install 'pip-tools @ file:./'
ERROR: Invalid requirement: 'pip-tools @ file:./'

$ pip install 'pip-tools @ file:.'
ERROR: Invalid requirement: 'pip-tools @ file:.'

$ pip install 'pip-tools @ file:pip-tools'
ERROR: Invalid requirement: 'pip-tools @ file:pip-tools'
```

FlorentJeannot replied:
Sorry about that, you're definitely right. There's even an example in pip's documentation about it: https://pip.pypa.io/en/stable/cli/pip_install/#requirement-specifiers

I have updated the code again. It looks simpler IMO now. Let me know what you think.
Also I'd love to add a test in `test_cli_compile.py` about a direct reference with a relative path in `file:` but it seems tricky to me because as far as I know the unit tests are working in a temp dir.

FlorentJeannot replied:
@atugushev I have updated the tests in `test_cli_compile.py`.

atugushev commented on tests/test_utils.py:
```diff
@@ -32,9 +32,54 @@ def test_format_requirement(from_line):
+            id="direct reference lowered case",
+        ),
+        pytest.param(
+            "exemple @ https://example.com/example.zip#egg=example",
```
```suggestion
            "example @ https://example.com/example.zip#egg=example",
```

Fixes a typo

FlorentJeannot replied:
Thanks for catching this. 👍

atugushev commented on piptools/utils.py:
```diff
@@ -115,7 +116,16 @@ def format_requirement(
+        # otherwise parse the URL
+        else:
+            parsed_url = urlparse(ireq.link.url)
+            if "egg=" in parsed_url.fragment:
```
```suggestion
            if ireq.link.egg_fragment:
```

`egg_fragment` can be used

FlorentJeannot replied:
Nice catch!

atugushev commented on piptools/utils.py:
```diff
@@ -115,7 +116,16 @@ def format_requirement(
     elif is_url_requirement(ireq):
-        line = ireq.link.url
+        # if the requirement has no name then it's just a URL
+        if not ireq.name:
```
For `if-else` statements I would recommend the following best practice:

> Put the case you normally expect to process first. This is in line with the general principle of putting code that results from a decision as close as possible to the decision.

```python
if ok:
    do_ok_stuff()
else:
    do_not_ok_stuff()
```

FlorentJeannot replied:
Thanks for your feedback, I've updated the logic.

atugushev reviewed (approved):
Thanks for addressing suggestions :+1:

---

As a side note, this transformation feels confusing:
```
example @ https://example.com/example.zip#egg=example
→
https://example.com/example.zip#egg=example
```
 
 I'd prefer the opposed instead:
```
https://example.com/example.zip#egg=example
→
example @ https://example.com/example.zip
```

AFAIK egg fragment appears to be [discouraged](https://github.com/pypa/pip/issues/1289#issuecomment-757402359) for direct references. However, pip is going to refactor this within https://github.com/pypa/pip/issues/1289, so it would make sense to address that later, as soon as the issue is resolved.

FlorentJeannot commented:
> Thanks for addressing suggestions 👍
> 
> As a side note, this transformation feels confusing:
> 
> ```
> example @ https://example.com/example.zip#egg=example
> →
> https://example.com/example.zip#egg=example
> ```
> 
> I'd prefer the opposed instead:
> 
> ```
> https://example.com/example.zip#egg=example
> →
> example @ https://example.com/example.zip
> ```
> 
> AFAIK egg fragment appears to be [discouraged](https://github.com/pypa/pip/issues/1289#issuecomment-757402359) for direct references. However, pip is going to refactor this within [pypa/pip#1289](https://github.com/pypa/pip/issues/1289), so it would make sense to address that later, as soon as the issue is resolved.

Hello @atugushev,

Thank you for your feedback and your approval for this PR.
I suggest to mention "PEP 508 support" in the `CHANGELOG.md` instead of PEP 440 as it seems to be more appropriate.

Regarding your side note, I agree with you and I'll happily provide a PR for this when https://github.com/pypa/pip/issues/1289 will be resolved.

atugushev commented:
@FlorentJeannot 

We should also support extras, see https://github.com/jazzband/pip-tools/issues/1450 for details. Would you like to address it here or in a following up PR?

FlorentJeannot commented:
> @FlorentJeannot
> 
> We should also support extras, see #1450 for details. Would you like to address it here or in a following up PR?

@atugushev 

I will do it in another PR since this one has been approved.

Also I have another question. It seems that `pip` already converts URLs to the PEP 508 format. See https://github.com/pypa/pip/blob/7da05f33345de4c59930398aefa011ed7ccdeca3/src/pip/_vendor/packaging/requirements.py#L148
Is there a good reason to wait for https://github.com/pypa/pip/issues/1289 or could we do it like them now?

atugushev commented:
@FlorentJeannot 

Thanks for the PR! Awesome job 👍🏻 

> Is there a good reason to wait for [pypa/pip#1289](https://github.com/pypa/pip/issues/1289) or could we do it like them now?

Could you elaborate on how code in [`Requirement.__str__`()](https://github.com/pypa/pip/blob/7da05f33345de4c59930398aefa011ed7ccdeca3/src/pip/_vendor/packaging/requirements.py#L148) related to egg fragment refactoring in pypa/pip#1289?

FlorentJeannot commented:
@atugushev Thanks 👍  Happy to contribute to this project 😄 

So I believe `Requirement.__str__()` is what is used to format the requirement in `pip`. I think that's the equivalent to the function I modified in `piptools` to support the direct reference.
In their code they seem to always convert requirements to the direct reference format when they have a URL with a name.

So for example if you do this:
`pip install 'git+https://github.com/encode/uvicorn.git@0.14.0#egg=uvicorn[standard]'`

Even though it was specified with an egg, the output of `pip freeze` will be a direct reference for `uvicorn` (which makes me wonder if we shouldn't do like them already):

```
asgiref==3.4.1
click==8.0.1
h11==0.12.0
httptools==0.2.0
pep517==0.10.0
python-dotenv==0.18.0
PyYAML==5.4.1
toml==0.10.2
uvicorn @ git+https://github.com/encode/uvicorn.git@87da6cf4082cd28306bdb78d7d42552d131cc179
uvloop==0.15.3
watchgod==0.7
websockets==9.1
```

atugushev commented:
@FlorentJeannot 

> if we shouldn't do like them already

Ideally, we should. Feel free to shoot a PR if think it's doable.

FlorentJeannot commented:
> @FlorentJeannot
> 
> > if we shouldn't do like them already
> 
> Ideally, we should. Feel free to shoot a PR if you think it's doable.

Will do!
